fishtank/src/cluster.ts

```diff
@@ -25,23 +33,68 @@ export class Cluster {
     return `${this.name}_${name}`
   }
 
-  async init(): Promise<void> {
-    return this.backend.createNetwork(this.networkName(), {
+  async init(options?: { bootstrap?: boolean | BootstrapOptions }): Promise<void> {
+    await this.backend.createNetwork(this.networkName(), {
       attachable: true,
       internal: true,
       labels: { [CLUSTER_LABEL]: this.name },
     })
+
+    if (typeof options?.bootstrap === 'undefined' || options?.bootstrap === true) {
+      return this.bootstrap()
+    } else if (typeof options?.bootstrap === 'object') {
+      return this.bootstrap(options?.bootstrap)
+    }
+  }
+
+  async bootstrap(options?: BootstrapOptions): Promise<void> {
+    await this.internalSpawn({
+      name: options?.nodeName ?? DEFAULT_BOOTSTRAP_NODE_NAME,
+      image: options?.nodeImage,
+      extraLabels: {
+        [NODE_ROLE_LABEL]: BOOTSTRAP_NODE_ROLE,
+      },
+    })
+  }
+
+  private async bootstrapNodes(): Promise<Node[]> {
+    return (
+      await this.backend.list({
+        labels: {
+          [CLUSTER_LABEL]: this.name,
+          [NODE_ROLE_LABEL]: BOOTSTRAP_NODE_ROLE,
+        },
+      })
+    ).map((container) => new Node(this, container.name.slice(this.name.length + 1)))
   }
 
   async spawn(options: { name: string; image?: string }): Promise<Node> {
+    const extraArgs = []
+    for (const bootstrapNode of await this.bootstrapNodes()) {
+      extraArgs.push('--bootstrap', bootstrapNode.name)
+    }
+    return this.internalSpawn({ extraArgs, ...options })
+  }
+
+  private async internalSpawn(options: {
+    name: string
+    image?: string
+    extraArgs?: string[]
+    extraLabels?: Labels
+  }): Promise<Node> {
     const containerName = this.containerName(options.name)
+    const args = ['start', ...(options.extraArgs ?? [])]
     await this.backend.runDetached(options.image ?? DEFAULT_IMAGE, {
+      args,
       name: containerName,
       networks: [this.networkName()],
       hostname: options.name,
-      labels: { [CLUSTER_LABEL]: this.name },
+      labels: {
+        [CLUSTER_LABEL]: this.name,
+        ...options.extraLabels,
+      },
     })
-    return new Node(this, containerName)
```

Comment:
why not use `containerName`

Reply:
Because nodes talk to each other by name, not by container name, so the name is an information that the `Node` class needs.

Reply:
What exact do you refer to for  `nodes talk`? What function do you refer to in Node class interface?

Reply:
The nodes communicate over TCP/IP in their isolated network. The Node class does not have to do anything for this to happen, but code surrounding it needs to have a way to know the node names to reference them at the DNS level. This is used on this line:

```typescript
extraArgs.push('--bootstrap', bootstrapNode.name)
```

Here `bootstrapNode.name` is that `name` that we're talking about. If this was a container name, the name wouldn't resolve and this whole mechanism wouldn't work.

Reply:
Thanks for the explanation